internal/update.go

```diff
@@ -0,0 +1,354 @@
+package internal
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+
+	"github.com/gogo/protobuf/types"
+	commonpb "go.temporal.io/api/common/v1"
+	protocolpb "go.temporal.io/api/protocol/v1"
+	updatepb "go.temporal.io/api/update/v1"
+	"go.temporal.io/sdk/converter"
+	"go.temporal.io/sdk/internal/protocol"
+)
+
+type updateState string
+
+const (
+	updateStateNew              updateState = "New"
+	updateStateRequestInitiated             = "RequestScheduled"
+	updateStateAccepted                     = "Accepted"
+	updateStateCompleted                    = "Completed"
+
+	updateProtocolV1 = "temporal.api.update.v1"
+)
+
+type (
+	// UpdateCallbacks supplies callbacks for the different stages of processing
+	// a workflow update.
+	UpdateCallbacks interface {
+		// Accept is called for an update after it has passed validation an
+		// before execution has started.
+		Accept()
+
+		// Reject is called for an update if validation fails.
+		Reject(err error)
+
+		// Complete is called for an update with the result of executing the
+		// update function. If the provided error is non-nil then the overall
+		// outcome is understood to be a failure.
+		Complete(success interface{}, err error)
+	}
+
+	// updateEnv encapsulates the utility functions needed by update protocol
+	// instance in order to implement the UpdateCallbacks interface. This
+	// interface is conveniently implemented by
+	// *workflowExecutionEventHandlerImpl.
+	updateEnv interface {
+		GetFailureConverter() converter.FailureConverter
+		GetDataConverter() converter.DataConverter
+		Send(*protocolpb.Message)
+	}
+
+	// updateProtocol wraps an updateEnv and some protocol metadata to
+	// implement the UpdateCallbacks abstraction. It handles callbacks by
+	// sending protocol lmessages.
+	updateProtocol struct {
+		protoInstanceID string
+		requestMsgID    string
+		requestSeqID    int64
+		initialRequest  updatepb.Request
+		scheduleUpdate  func(name string, args *commonpb.Payloads, header *commonpb.Header, callbacks UpdateCallbacks)
+		env             updateEnv
+		state           updateState
+	}
+
+	// updateHandler is the underlying type that is registered into a workflow
+	// environment when the user-code in a workflow registers an update callback
+	// for a given name. It offers the ability to invoke the associated
+	// execution and validation functions.
+	updateHandler struct {
+		fn         interface{}
+		validateFn interface{}
+		name       string
+	}
+)
+
+// newUpdateResponder constructs an updateProtocolResponder instance to handle
+// update callbacks.
+func newUpdateProtocol(
+	protoInstanceID string,
+	scheduleUpdate func(name string, args *commonpb.Payloads, header *commonpb.Header, callbacks UpdateCallbacks),
+	env updateEnv) *updateProtocol {
```

Comment:
```suggestion
	env updateEnv,
) *updateProtocol {
```

A bit clearer IMO

Reply:
Agreed. Not committing yours only because this whole thing is going to get wiped when the new upstream API comes in.